pgdog/src/backend/pool/dns_cache.rs

```diff
@@ -0,0 +1,388 @@
+use futures::future::join_all;
+use once_cell::sync::Lazy;
+use std::collections::{HashMap, HashSet};
+use std::net::IpAddr;
+use std::sync::{Arc, RwLock};
+use std::time::{Duration, Instant};
+use tokio::net::lookup_host;
+use tokio::select;
+use tokio::time::sleep;
+use tokio_util::sync::CancellationToken;
+use tracing::{debug, error, info, warn};
+
+// -------------------------------------------------------------------------------------------------
+// ------ Constant(s) ------------------------------------------------------------------------------
+
+const DEFAULT_TTL: Duration = Duration::from_secs(30);
+
+// -------------------------------------------------------------------------------------------------
+// ------ Singleton(s) -----------------------------------------------------------------------------
+
+static DNS_CACHE: Lazy<Arc<DnsCache>> = Lazy::new(|| {
+    Arc::new(DnsCache {
+        cache: Arc::new(RwLock::new(HashMap::new())),
+        hostnames: Arc::new(RwLock::new(HashSet::new())),
+        ttl: Arc::new(RwLock::new(DEFAULT_TTL)),
```

Comment:
You can fetch this from the config and it will always reflect the latest version of it:

```rust
crate::config::config().config.general.dns_ttl
```